File: bob-client/app/src/main/java/fr/dmconcept/bob/client/communications/SendStepsAsyncTask.java
package fr.dmconcept.bob.client.communications;

import android.os.AsyncTask;
import android.util.Log;

import org.apache.http.HttpResponse;
import org.apache.http.HttpStatus;
import org.apache.http.client.HttpClient;
import org.apache.http.client.methods.HttpPost;
import org.apache.http.impl.client.DefaultHttpClient;

import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.Reader;

import fr.dmconcept.bob.client.models.serializers.ProjectStepSerializer;
import fr.dmconcept.bob.client.models.serializers.ServoConfigSerializer;

class SendStepsAsyncTask extends AsyncTask<SendStepInput, Void, SendStepResult> {

    private static final String TAG = "communication.SendSteps";

    private static final String RESPONSE_OK = "BOB";

    private String mHost;

    private int mPort;

    public SendStepsAsyncTask(String host, int port) {
        this.mHost = host;
        this.mPort = port;
    }

    @Override
    protected SendStepResult doInBackground(SendStepInput... params) {

        SendStepResult result = new SendStepResult();

        try {

            Log.i(TAG, "doInBackground() - Sending request...");
            sendRequest(params[0]);
            Log.i(TAG, "doInBackground() - Request sent successfully");

        } catch (IOException e) {

            result.setError(e.getMessage());

            Log.i(TAG, "doInBackground() - Request error: " + e.getMessage());
        }

        return result;

    }

    private void sendRequest(SendStepInput input) throws IOException {

        Log.i(TAG, "sendRequest()");

        String serializedServoConfigs = ServoConfigSerializer.serialize(input.getBoardConfig().getServoConfigs());
        String serializedSteps        = ProjectStepSerializer.serialize(input.getSteps());

        String url = "http://" + mHost + ":" + mPort + "/step?servoconfig=" + serializedServoConfigs + "&steps=" + serializedSteps;

        Log.i(TAG, "sendRequest() -> " + url);

        HttpClient httpClient = new DefaultHttpClient();
        HttpPost httpPost     = new HttpPost(url);

        HttpResponse response = httpClient.execute(httpPost);

        int status = response.getStatusLine().getStatusCode();

        if (status != HttpStatus.SC_OK) {
            Log.d(TAG, "sendRequest() <- HTTP " + status);
            throw new RuntimeException("HTTP " + status);
        }

        String body = readIt(response.getEntity().getContent());

        Log.d(TAG, "sendRequest() <- HTTP " + status + " - " + body);

        if (!body.equals(RESPONSE_OK))
            throw new RuntimeException("Invalid response: " + body);

    }

    // Reads an InputStream and converts it to a String.
    private String readIt(InputStream stream) throws IOException {
        Reader reader;
        reader = new InputStreamReader(stream, "UTF-8");
        char[] buffer = new char[1024];
        reader.read(buffer);
        return new String(buffer);
    }

}

File: bob-client/src/main/scala/fr/dmconcept/bob/client/communications/SendStepsAsyncTask.scala
package fr.dmconcept.bob.client.communications

import SendStepsAsyncTask._
import android.os.AsyncTask
import android.util.Log
import fr.dmconcept.bob.client.communications.SendStepsAsyncTask.{SendStepResult, SendStepInput}
import fr.dmconcept.bob.client.models.json.BobJsonProtocol._
import fr.dmconcept.bob.client.models.{Step, BoardConfig}
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader
import org.apache.http.HttpResponse
import org.apache.http.HttpStatus
import org.apache.http.client.methods.HttpPost
import org.apache.http.impl.client.DefaultHttpClient
import spray.json._


object SendStepsAsyncTask {

  final val RESPONSE_OK = "BOB"

  def log(message: String) = Log.i("communication.SendSteps", message)

  case class SendStepInput(
    boardConfig : BoardConfig,
    steps       : Vector[Step]
  )

  case class SendStepResult(
    error: Option[String]
  )

}

case class SendStepsAsyncTask(

  host: String,

  port: Int

) extends AsyncTask[SendStepInput, Void, SendStepResult] {

  override def doInBackground(params: SendStepInput*): SendStepResult = {

    try {

        log("doInBackground() - Sending request...")
        sendRequest(params(0))
        log("doInBackground() - Request sent")

        SendStepResult(None)

    } catch {
      case e: Throwable =>
        log("doInBackground() - Request error: ${e.getMessage()}")
        SendStepResult(Option(e.getMessage))
    }

  }

  @throws [IOException]("if the network fails")
  def sendRequest(input: SendStepInput) {

    val serializedServoConfigs: String = input.boardConfig.toJson.compactPrint
    val serializedSteps       : String = input.steps.toJson.compactPrint

    //TODO put request in body
    val url = s"http://$host:$port/step?servoconfig=$serializedServoConfigs&steps=$serializedSteps"

    log(s"sendRequest($url)")

    val httpClient = new DefaultHttpClient()
    val httpPost   = new HttpPost(url)

    val response = httpClient.execute(httpPost)

    response.getStatusLine.getStatusCode match {

      case HttpStatus.SC_OK =>

        // TODO use scala iostream
        def readLine(response: HttpResponse) = new BufferedReader(
            new InputStreamReader(response.getEntity.getContent, "UTF-8")
          ).readLine

        val body = readLine(response)

        log(s"sendRequest() <- HTTP 200 - $body")

        if (!body.equals(RESPONSE_OK))
          throw new RuntimeException("Invalid response: $body")

      case status: Int =>
        throw new RuntimeException(s"HTTP $status")

    }

  }

}
